content/en/docs/reference/using-api/deprecation-policy.md

```diff
@@ -73,21 +73,19 @@ knows how to convert between them in both directions.  Additionally, any new
 field added in v2 must be able to round-trip to v1 and back, which means v1
 might have to add an equivalent field or represent it as an annotation.
 
-**Rule #3: An API version in a given track may not be deprecated until a new
-API version at least as stable is released.**
+**Rule #3: An API version in a given track may not be deprecated in favor of a less stable API version.**
 
-GA API versions can replace GA API versions as well as beta and alpha API
-versions.  Beta API versions *may not* replace GA API versions.
+  * GA API versions can replace beta and alpha API versions.
+  * Beta API versions can replace earlier beta and alpha API versions, but *may not* replace GA API versions.
+  * Alpha API versions can replace earlier alpha API versions, but *may not* replace GA or beta API versions.
 
-**Rule #4a: Other than the most recent API versions in each track, older API
-versions must be supported after their announced deprecation for a duration of
-no less than:**
+**Rule #4a: minimum API lifetime is determined by the API stability level**
 
-   * **GA: 12 months or 3 releases (whichever is longer)**
-   * **Beta: 9 months or 3 releases (whichever is longer)**
-   * **Alpha: 0 releases**
+   * **GA: may be marked as deprecated, but must not be removed**
```

Comment:
in the [mailing list thread](https://groups.google.com/g/kubernetes-sig-architecture/c/fdnT600f9Fs/m/zWyPJbV6DAAJ), you suggested referencing a Kubernetes major version (e.g. Kubernetes v2.xxx) as a point at which stable APIs could be dropped. My concerns with language like that are:
* there are no current plans for a v2.x stream
* I wouldn't want this to serve as a motiviation for a v2 that drops stable APIs (I fear a python2/python3-style split of the ecosystem)

I'm curious what others think here

Reply:
i tend to think that any kubernetes 2.x is going to be a layer around, above, or alongside kube v1.  I highly doubt that Kube as a project will break continuity of stable API in such a dramatic way.  Linux syscall has been fairly stable for 10+ years, and I tend to view our core stable APIs in the same vein.

I'd probably lean more towards creating the space for people to build their own v2 abstractions AND offer side by side alternatives to v1 (endpoint to endpoint slice is such an example) because that preserves the momentum of the project.  Jordan's python2/3 point I think is the relevant cautionary tale - if someone wants to build to kube v2 AND decides turns off v1, that's great.  But that won't necessarily be our focus and we will not prioritize that.

Reply:
Rather than fork Kubernetes, I'd like to see the 1.x line of minor releases end (still with patch releases), and a new v2.x series start. What I'm imagining is that the last v1.x release and the v2.x release would be really similar, but the shift to v2.x would allow for some (well-publicised) removals.

So it'd be a new major version that looked a lot like the previous release.

As an example of what I'd turn off in a putative Kubernetes v2
- `componentstatuses` API endpoints (this was the one I had in mind with my reservation about the change)
   - Gone, farewell, not served, not in the OpenAPI.
   - `kubectl` etc deny all knowledge
   - An error if you try to access them anyway

Leaving things like this as part of Kubernetes., even if they are deprecated, means some amount of maintenance cost and documentation effort _forever_. That's not so bad if it's only ever ComponentStatus but can we be sure there'll never be another API we fully move away from?

As an example of what I, personally, wouldn't want to turn off in a putative Kubernetes v2
- `endpoints` API endpoints (this was the one I had in mind with my reservation about the change)
   - Widely used at the moment; I don't relish the deprecation / migration work
   - Handling beta API deprecations and removal is already hard enough!

I am neutral on whether or not we remove, say, HorizontalPodAutoscaler v1 now that a new v2 API is stable. I'm sure other people have a better perspective on that/


If removing ComponentStatus splits the ecosystem, I'd be shocked. The overall idea is to drop the APIs that people have already actually stopped using.

PS. We could, exceptionally, make a commitment to support the final v1.x minor release for a longer period, say 18 months.

Reply:
i wonder if k8s, long term would be comparable to the kernel ABIs stability. the surface there is much smaller, lower level and with less user facing exposure due to wrapping around system libraries. while some of the k8s stable APIs are directly user facing and much higher level.

the python fragmentation story is something i don't see happening in a mostly corporate driven development model as in k8s. a company choosing to remain on an old version of k8s while there rest of the project moves forward would not be a good business decision.

i think etcd is an example where they released a new incompatible API layer v3 (incompatible with v2) and that version of the etcd v3.x.x only works with the new layer. then the project established guides for migrating users to v3. if k8s ends up making major changes in v2.x.x, that new version may or may not be desirably compatible with the v1...but according to the policy - it must be.

i think it's worth leaving a small "backdoor" in the language here that this rule would apply to the current state of k8s versioning...if k8s moves to v2.x or has an unforeseen significant redesign, k8s v1.x would continue supporting the old set of stable APIs, but v2.x might not.

Reply:
> componentstatuses API endpoints (this was the one I had in mind with my reservation about the change)

Ultimately the point made in the email thread is that no matter what, people contributing to Kube have to spend time dealing with implications of removing this API.  Creating a v2 version of Kube, debating what is part of v2, writing blog posts, calming panicked users, setting up build infrastructure, having the release team deal with 2 completely separate streams is a hugely expensive way of removing the API.  The default option is always "do nothing" and that is cheaper than any other approach.  The next cheapest is adding a selective disablement flag.  Once you get past that, the cost goes up significantly.

I really think the best way to measure these sorts of deprecations is focusing on the benefit you get out of it in measurable units, and imagine spending those measurable units on something you personally care about even more (like the Kubelet being 5% more efficient or a new flag on stateful sets).  

> if k8s moves to v2.x or has an unforeseen significant redesign, k8s v1.x would continue supporting the old set of stable APIs, but v2.x might not.

I think it's reasonable to describe this effort as applying to k8s v1.x (because we have adopted semantic versioning, and have implicitly chosen to honor that mindset via no breakages to stable APIs).  If we were to consider a k8s v2, we would need to revisit these topics anyway.

Reply:
Just to clarify one point, I don't want to advocate for an immediate v2.x release; I do only want to leave room for it to happen at some future point, when an API that's no longer used is hurting the project enough to justify that. (I don't think ComponentStatus is that API, so maybe it was a misleading example).

Reply:
I agree with @sftim - if we wrap  all of these rules with "within a major version of Kubernetes", we at least  leave room to discuss a v2 and whether it would be worth the effort to EOL some APIs.  I don't even mind saying "as of Q1 2022, there are no plans for a v2.x.x of Kubernetes"

Reply:
bounding this to v1.x and clearly stating there aren't current plans for a v2.x seems reasonable

Reply:
added a commit to that effect